test/js/lib/validate-file.test.js

```diff
@@ -45,20 +48,99 @@ describe('Client-side file validation', () => {
+    // eslint-disable-next-line max-len
+    // Mirrors https://github.com/broadinstitute/scp-ingest-pipeline/blob/af1c124993f4a3e953debd5a594124f1ac52eee7/tests/test_annotations.py#L56
+    mockReadLinesAndType('dup_headers_v2.0.0.tsv')
+    const { errors, summary } = await validateFile({}, 'metadata')
+    expect(errors).toHaveLength(1)
+    expect(summary).toBe('Your metadata file had 1 error')
+  })
+
+  it('reports no error with good cluster CSV file', async () => {
+    // Confirms no false positive due to comma-separated values
+    mockReadLinesAndType('cluster_comma_delimited.csv')
+    const { errors } = await validateFile({}, 'cluster')
+    expect(errors).toHaveLength(0)
+  })
+
+  it('catches mismatched header counts', async () => {
+    mockReadLinesAndType('header_count_mismatch.tsv')
+    const { errors, summary } = await validateFile({}, 'metadata')
+    expect(errors).toHaveLength(1)
+    expect(summary).toBe('Your metadata file had 1 error')
+  })
+
+  it('catches multiple header errors', async () => {
+    // eslint-disable-next-line max-len
+    // Mirrors https://github.com/broadinstitute/scp-ingest-pipeline/blob/af1c124993f4a3e953debd5a594124f1ac52eee7/tests/test_annotations.py#L112
+    mockReadLinesAndType('error_headers_v2.0.0.tsv')
+    const { errors, summary } = await validateFile({}, 'metadata')
+    expect(errors).toHaveLength(3)
+    expect(summary).toBe('Your metadata file had 3 errors')
+  })
+
+  it('fails when no coordinates in cluster file', async () => {
+    // Confirms this validation does not report false negatives
+    //
+    // eslint-disable-next-line max-len
+    // Mirrors https://github.com/broadinstitute/scp-ingest-pipeline/blob/af1c124993f4a3e953debd5a594124f1ac52eee7/tests/test_cluster.py#L9
+    mockReadLinesAndType('cluster_bad_no_coordinates.txt')
+    const { errors } = await validateFile({}, 'cluster')
+    expect(errors).toHaveLength(1)
+  })
+
+  it('passes when no coordinates in cluster file', async () => {
+    // Confirms this validation does not report false positive
+    //
+    // eslint-disable-next-line max-len
+    // Mirrors https://github.com/broadinstitute/scp-ingest-pipeline/blob/af1c124993f4a3e953debd5a594124f1ac52eee7/tests/test_cluster.py#L21
+    mockReadLinesAndType('cluster_example.txt')
+    const { errors } = await validateFile({}, 'cluster')
+    expect(errors).toHaveLength(0)
+  })
+
+  it('fails when coordinates in metadata file', async () => {
+    // Confirms this validation does not report false negatives
+    //
+    // eslint-disable-next-line max-len
+    // Mirrors https://github.com/broadinstitute/scp-ingest-pipeline/blob/af1c124993f4a3e953debd5a594124f1ac52eee7/tests/test_cell_metadata.py#L17
+    mockReadLinesAndType('metadata_bad_has_coordinates.txt')
+    const { errors } = await validateFile({}, 'metadata')
+    expect(errors).toHaveLength(1)
+  })
+
+  it('passes when no coordinates in metadata file', async () => {
```

Comment:
I really appreciate the two positive tests in this suite!

Reply:
Thanks is really due to @knapii-developments and @jlchang, whose bullet-proof tests I translate and refine here.